assets/js/base/hooks/use-store-cart-coupons.js

```diff
@@ -21,20 +23,80 @@ import { useStoreCart } from './use-store-cart';
  */
 export const useStoreCartCoupons = () => {
 	const { cartCoupons, cartIsLoading } = useStoreCart();
+	const {
+		addErrorNotice,
+		addSuccessNotice,
+		addInfoNotice,
+	} = useStoreNotices();
 
-	const results = useSelect( ( select, { dispatch } ) => {
-		const store = select( storeKey );
-		const isApplyingCoupon = store.isApplyingCoupon();
-		const isRemovingCoupon = store.isRemovingCoupon();
-		const { applyCoupon, removeCoupon } = dispatch( storeKey );
-
-		return {
-			applyCoupon,
-			removeCoupon,
-			isApplyingCoupon,
-			isRemovingCoupon,
-		};
-	}, [] );
+	const results = useSelect(
+		( select, { dispatch } ) => {
+			const store = select( storeKey );
+			const isApplyingCoupon = store.isApplyingCoupon();
+			const isRemovingCoupon = store.isRemovingCoupon();
+			const { applyCoupon, removeCoupon } = dispatch( storeKey );
+
+			const applyCouponWithNotices = ( couponCode ) => {
+				applyCoupon( couponCode )
+					.then( ( result ) => {
+						if ( result === true ) {
+							addSuccessNotice(
+								sprintf(
+									// translators: %s coupon code.
+									__(
+										'Coupon code "%s" has been applied to your cart',
+										'woo-gutenberg-products-block'
+									),
+									couponCode
+								),
+								{
+									id: 'coupon-form',
+								}
+							);
+						}
+					} )
+					.catch( ( error ) => {
+						addErrorNotice( error.message, {
+							id: 'coupon-form',
+						} );
+					} );
```

Comment:
I'm fine with leaving this for now (and great job 👏 ), but it relies on actionCreators always [returning a promise](https://github.com/WordPress/gutenberg/blob/11963546b7921f26d6f21a7b74e27ddfb9b8e1f4/packages/data/src/namespace-store/index.js#L206-L208) that resolves to the action result.  There was some discussion around this behaviour in GB core at one point but it looks like that got resolved by a pull (I did!) [done here](https://github.com/WordPress/gutenberg/pull/14830).  For some reason, I thought that we were eventually going to revert that but looks like it's okay.  Worth noting that the discussion [here](https://github.com/WordPress/gutenberg/issues/17485), might have some impact.

Reply:
I'll defer to your advice here @nerrad. The alternative was having these dispatch calls in the `applyCoupon` action itself (resolver.js) but it felt very opinionated to throw real notices there rather than via this hook. Any other/better suggestions I'm open.

Reply:
Let's leave it, even considering the unresolved discussion I don't think it will affect this implementation.